src/librustc_traits/lowering.rs

```diff
@@ -99,6 +101,38 @@ impl<'tcx> Lower<Goal<'tcx>> for ty::Predicate<'tcx> {
     }
 }
 
+/// For a given `where_clause`, returns a clause `FromEnv(WC) :- Implemented(Self: Trait<P1..Pn>)`.
```

Comment:
As an aside, to avoid writing again code about binders, I'd rather see something like:
```rust
impl<'tcx> DomainGoal<'tcx> {
    fn into_from_env_goal(self) -> DomainGoal<'tcx> {
        match self {
            DomainGoal::Holds(wca) => DomainGoal::FromEnv(wca),
            other => other // unchanged
        }
    }
}
```
then the same thing recursively for `Goal`:
```rust
impl<'tcx> Goal<'tcx> {
    fn into_from_env_goal(self) -> Goal<'tcx> {
        match self {
            Goal::DomainGoal(dg) => Goal::DomainGoal(dg.into_from_env_goal()),
            Goal::Quantified(q, g) => Goal::Quantified(q, g.into_from_env_goal()),
            /* ... */
        }
    }
}
```

and then you can just do something like:
```rust
let predicates = tcx.predicates_of(def_id).predicates.lower();

/* do appropriate things to remove the `Self: Trait` predicate */

let predicates = predicates.into_iter().map(|p| p.into_from_env_goal());
````

Reply:
Actually I was mistaken this would not quite work since we only can have `DomainGoal` as a consequence. Ok so do as you want 😀

Reply:
Anyway I think I took a slightly wrong approach when initially translating HRTBs. I'll issue a fix which should benefit the ergonomics of your PR.

Reply:
There might be a way to move the `Lower` types around a bit to do what we want...